be1-go/popcha/server.go

```diff
@@ -0,0 +1,410 @@
+package popcha
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"github.com/aaronarduino/goqrsvg"
+	"github.com/ajstarks/svgo"
+	"github.com/boombuler/barcode/qr"
+	"github.com/gorilla/mux"
+	"github.com/rs/zerolog"
+	"github.com/zitadel/oidc/v2/pkg/oidc"
+	"github.com/zitadel/oidc/v2/pkg/op"
+	"golang.org/x/xerrors"
+	"html/template"
+	"net/http"
+	"os"
+	"popstellar/hub"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	// formatting for auth request validation error
+	errValidAuthFormat = "Error while validating the auth request: %s"
+
+	//error message for invalid response type
+	errInvalidRepsonseType = "response type is invalid for the implicit flow"
+
+	//error message for unimplemented methods of op.Client
+	errUnimplementedMethod = "this method is not implemented for our protocol"
+
+	// defaultValue for the ID token lifetime (in hours)
+	tokenLifeTimeHour = 24
+
+	// qrCodeWebPage file path for valid QRCode Displaying page
+	qrCodeWebPage = "qrcode/popcha.html"
+
+	// qrSize size of QRCode SVG
+	qrSize = 10
+
+	badRequestCode = 400
+)
+
+// constant parameter names
+const (
+	ClientID     = "client_id"
+	Nonce        = "nonce"
+	Scope        = "scope"
+	RedirectURI  = "redirect_uri"
+	ResponseType = "response_type"
+	State        = "state"
+	ResTypeMulti = "id_token token"
+	OpenID       = "openid"
+	Profile      = "profile"
+	LoginHint    = "login_hint"
+)
+
+// clientParams implements op.Client
+// as PoPCHA don't support registered clients, clientParams is an alternative which defines  parameters
+// for any client sending an auth request. As an example, any client requesting authorization has to use
+// the implicit flow.
+type clientParams struct {
+	clientID     string
+	redirectURIs []string
+	resType      oidc.ResponseType
+}
+
+/*
+Most of the methods of op.Client are implemented, however some of them
```

Comment:
Small suggestion: I think that physically separating the functions that are implemented and the ones that are not by some comments (in two different "blocs") would make the code more readable. Future programmers might find it useful to only see the functions they can directly use and the ones that just panic separately.

Reply:
True! That's a very good suggestion definitely.